Document class and module attributes

daniel-zullo-frequenz opened the pull request:
Add or amend docstrings for all classes and modules where needed.

Fixes #74

daniel-zullo-frequenz committed:
Document class and module attributes

Add or amend docstrings for all classes and modules
where needed.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Rename variable according to the naming convention

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Remove type variable no longer needed

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Fix typos in code documentation

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

llucax commented on src/frequenz/sdk/_api_client/api_client.py:
```diff
@@ -11,8 +11,13 @@ class ApiProtocol(Enum):
     """Enumerated values of supported API types."""
 
     GRPC = 1
+    """gRPC API."""
+
     REST = 2
+    """REST API."""
+
     FILESYSTEM = 3
+    """Filesystem API."""
```
I wonder what is this about, Filesystem API? But anyway, not the scope of this pr...

llucax replied:
This is old should probably be removed in the future.

llucax commented on src/frequenz/sdk/_internal/_singleton_meta.py:
```diff
@@ -11,7 +11,10 @@ class SingletonMeta(type):
     """This is a thread-safe implementation of Singleton."""
 
     _instances: Dict[Any, type] = {}
+    """A dictionary of instances of the singleton classes."""
```
In general for attributes I think the pattern "The ..." instead of "A ..." works better. In new code we usually use that, so I would try to create new docs following this style.

daniel-zullo-frequenz replied:
will update this case and similar places where is needed

llucax commented on src/frequenz/sdk/_internal/_singleton_meta.py:
```diff
@@ -11,7 +11,10 @@ class SingletonMeta(type):
+    """A dictionary of instances of the singleton classes."""
+
     _lock: Lock = Lock()
+    """A lock to ensure thread safety."""
```
It might be good to mention here what is this lock protecting.

llucax commented on src/frequenz/sdk/actor/power_distributing/_battery_status.py:
```diff
@@ -81,15 +78,21 @@ class _ComponentStreamStatus:
 @dataclass
 class _BlockingStatus:
     min_duration_sec: float
+    """Minimum blocking duration (in seconds)."""
```
Here too, I would use "The minimum..."

llucax commented on src/frequenz/sdk/actor/power_distributing/_battery_status.py:
```diff
@@ -81,15 +78,21 @@ class _ComponentStreamStatus:
             f"than maximum blocking duration ({self.max_duration_sec})"
         )
         self.last_blocking_duration_sec: float = self.min_duration_sec
+        """Last blocking duration (in seconds)."""  # pylint: disable=pointless-string-statement
```
Ouch, weird, I never seen this pylint check triggered before because of this. Do you have any idea why it is triggered here?

Oh, probably because it is in `__post_init__` instead of `__init__`?

Can't we define it in the regular attributes?

daniel-zullo-frequenz replied:
> Oh, probably because it is in __post_init__ instead of __init__?

yes, that was the reason.

> Can't we define it in the regular attributes?

Yes, we can but with a bit of refactoring as the attributes in __post_init__ are not passed at the class instantiation. I think I can remove one of the attributes that depends on another attribute of the class and for the other one just set a default value.

llucax replied:
OK, I see. I think the new code is better, and in general, when more advanced uses like these are needed I think it is better to just write a regular class instead.

llucax commented on src/frequenz/sdk/actor/power_distributing/_battery_status.py:
```diff
@@ -150,21 +153,25 @@ class BatteryStatusTracker:
     _battery_valid_relay: Set[BatteryRelayState.ValueType] = {
         BatteryRelayState.RELAY_STATE_CLOSED
     }
+    """Valid relay states for the battery."""
```
Is this representing one battery? If so I would also use "for **this** battery" or "for the battery we are tracking" to make it more clear about which battery you are talking about.

Same for the other below.

daniel-zullo-frequenz replied:
Actually they are the valid states that any battery can have. The same applies for the inverter. 
So I think `Valid relay states of the battery.` is the correct one

llucax replied:
Then I would say "of **a** battery", IMHO **the** suggests you are talking about one specific battery.

llucax replied:
Also I would make it more verbose:

```
The list of valid states of a yyy.

A yyy is in any other state will be considered as failing and log a warning.
```

Or similar.

shsms replied:
I think `the` makes sense as well, because this is not an enum, but a struct and there will be one instance for each battery.

daniel-zullo-frequenz replied:
Updated as suggested with slight modifications

llucax commented on src/frequenz/sdk/actor/power_distributing/result.py:
```diff
@@ -84,9 +84,16 @@ class PowerBounds:
     """Inclusion and exclusion power bounds for requested batteries."""
 
     inclusion_lower: float
+    """The inclusion power lower bound for the requested batteries."""
```
This sounds a bit funny, but I couldn't come up with a phrasing that is more clear, but mentioning it in case someone can think of something better 😁

daniel-zullo-frequenz replied:
@shsms do you have any suggestion here?

shsms replied:
5 options:

The lower {value/end/limit} of the inclusion power bounds ...
The lower bound of the {range/span} where power requests will be accepted ...

daniel-zullo-frequenz replied:
I think the latter couple of suggestions are a bit tricky as they need to explain the concept of inclusion/exclusion bounds.
So I'll go for `The lower value of the inclusion power bounds for the requested batteries` if there is no objections.

shsms replied:
that's how they're described in the battery pool,  and in the microgrid api: https://github.com/frequenz-floss/frequenz-sdk-python/blob/a8f3d62e93d226278c71fc650033fd00758d62a1/src/frequenz/sdk/timeseries/battery_pool/_result_types.py#L35

daniel-zullo-frequenz replied:
Exactly. Here it's just the internal representation of the power bounds so it makes sense to me to only explain it in terms of bounds/ranges rather than the inclusion/exclusion concept

llucax commented on src/frequenz/sdk/config/_config.py:
```diff
@@ -12,6 +12,7 @@
 _logger = logging.getLogger(__name__)
 
 T = TypeVar("T")
+"""Type variable for validating config values."""
```
We can keep this but in general I wouldn't document `TypeVar`s unless there is something interesting to say about them, also ingeneral they should be private.

daniel-zullo-frequenz replied:
Yes, they should be `private`. At the moment we have a mix of `_T` and `T`

llucax commented on src/frequenz/sdk/microgrid/client/_client.py:
```diff
@@ -41,8 +41,8 @@
 
-# Default timeout applied to all gRPC calls
 DEFAULT_GRPC_CALL_TIMEOUT = 60.0
+"""Default timeout for gRPC calls in seconds made by this client."""
```
```suggestion
"""Default timeout for gRPC calls made by this client (in seconds)."""
```
?

llucax commented on src/frequenz/sdk/microgrid/client/_connection.py:
```diff
@@ -10,7 +10,10 @@ class Connection(NamedTuple):
     """Metadata for a connection between microgrid components."""
 
     start: int
+    """The start index of the connection."""
+
     end: int
+    """The end index of the connection."""
```
What is a start and end index of a connection? It doesn't seem very clear to me at least...

daniel-zullo-frequenz replied:
True, I'll clarify it

llucax commented on src/frequenz/sdk/microgrid/component/_component.py:
```diff
@@ -67,16 +74,32 @@ class ComponentCategory(Enum):
+
     EV_CHARGER = components_pb.ComponentCategory.COMPONENT_CATEGORY_EV_CHARGER
+    """EV charger component."""
+
     CHP = components_pb.ComponentCategory.COMPONENT_CATEGORY_CHP
+    """CHP component."""
 
-    # types not yet supported by the API but which can be inferred
-    # from available graph info
     PV_ARRAY = 1000001
+    """PV array component.
```
I wonder if in the documentation we should expand acronyms like EV, CHP, PV, or if we should rely on the glossary...

Opinions @matthias-wende-frequenz @shsms?

shsms replied:
No harm in describing them in detail here, but these are low level interfaces, and we don't expect people who aren't familiar with these components to come and mess around here.  

But maybe we can mention the full forms in the high level interfaces in the logical meter, for `chp*_power` and `pv*_power`

llucax commented on src/frequenz/sdk/microgrid/component/_component.py:
```diff
@@ -67,16 +74,32 @@ class ComponentCategory(Enum):
     PV_ARRAY = 1000001
+    """PV array component.
+
+    Types not yet supported by the API but which can be inferred from available
+    graph info.
```
This sounds like an implementation detail that should be kept as a code comment rather than a docstring. @shsms ?

shsms replied:
I don't think this `PV_ARRAY` component should remain here at all.  We don't use it anymore.  This is from the old `MicrogridData` days.

daniel-zullo-frequenz replied:
will revert it

daniel-zullo-frequenz replied:
> I don't think this PV_ARRAY component should remain here at all. We don't use it anymore. This is from the old MicrogridData days.

Then I'll add a new commit to remove it if that's the case

daniel-zullo-frequenz replied:
I've removed PV_ARRAY. Would this need an entry into the release notes?

llucax commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_formula_generator.py:
```diff
@@ -31,6 +31,7 @@ class ComponentNotFound(FormulaGenerationError):
 
 
 NON_EXISTING_COMPONENT_ID = sys.maxsize
+"""A component ID that does not exist in the component graph."""
```
Maybe add some extra info? when this should be used?

daniel-zullo-frequenz replied:
Added extra info. @llucax please review the recently added documentation

llucax commented on src/frequenz/sdk/timeseries/_formula_engine/_tokenizer.py:
```diff
@@ -79,16 +79,24 @@ class TokenType(Enum):
+    """The type of the token."""
+
     value: str
+    """The value of the token."""
```
Maybe "The value associated to the token"?

daniel-zullo-frequenz replied:
```suggestion
    """The value associated to the token."""
```

llucax commented on src/frequenz/sdk/timeseries/_ringbuffer/buffer.py:
```diff
@@ -18,6 +18,7 @@
 from .._quantities import QuantityT
 
 FloatArray = TypeVar("FloatArray", List[float], npt.NDArray[np.float64])
+"""Type hint for the buffer container."""
```
Maybe use "Type variable" to be consistent with all the others you documented?

daniel-zullo-frequenz replied:
```suggestion
"""Type variable of the buffer container."""
```

llucax commented on src/frequenz/sdk/timeseries/battery_pool/_metric_calculator.py:
```diff
@@ -61,6 +63,7 @@ def battery_inverter_mapping(batteries: Iterable[int]) -> dict[int, int]:
 # Formula output types class have no common interface
 # Print all possible types here.
 T = TypeVar("T", Sample[Percentage], Sample[Energy], PowerMetrics, Sample[Temperature])
+"""Type of the formula output."""
```
Type variable?

daniel-zullo-frequenz replied:
```suggestion
"""Type variable of the formula output."""
```

llucax commented on src/frequenz/sdk/timeseries/ev_charger_pool/_state_tracker.py:
```diff
@@ -25,12 +25,18 @@ class EVChargerState(Enum):
     """State of individual ev charger."""
 
     UNSPECIFIED = "UNSPECIFIED"
+    """State is unspecified."""
```
```suggestion
    """State is unspecified."""

```

llucax commented on src/frequenz/sdk/timeseries/ev_charger_pool/_state_tracker.py:
```diff
@@ -25,12 +25,18 @@ class EVChargerState(Enum):
+    """EV Charger is missing."""
 
     IDLE = "IDLE"
+    """EV Charger is idle."""
```
```suggestion
    """EV Charger is idle."""

```

llucax commented on src/frequenz/sdk/timeseries/ev_charger_pool/_state_tracker.py:
```diff
@@ -25,12 +25,18 @@ class EVChargerState(Enum):
     IDLE = "IDLE"
+    """EV Charger is idle."""
     EV_PLUGGED = "EV_PLUGGED"
+    """EV Charger has an EV plugged in."""
```
```suggestion
    """EV Charger has an EV plugged in."""

```

llucax commented on src/frequenz/sdk/timeseries/ev_charger_pool/_state_tracker.py:
```diff
@@ -25,12 +25,18 @@ class EVChargerState(Enum):
     EV_PLUGGED = "EV_PLUGGED"
+    """EV Charger has an EV plugged in."""
     EV_LOCKED = "EV_LOCKED"
+    """EV Charger has an EV plugged in and locked."""
```
```suggestion
    """EV Charger has an EV plugged in and locked."""

```

llucax reviewed (commented):
A few comments. Of course I didn't check if all missing attributes, etc. were documented. 😁

daniel-zullo-frequenz committed:
Define regular attributes with default values

Pylint was complaining about documentation (pointless-string-statement)
for class attributes defined in the `__post_init__` method which
is not the recommended way to define attributes.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

llucax commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -146,9 +146,16 @@ def __init__(
         self.distribution_algorithm = DistributionAlgorithm(
             self.power_distributor_exponent
         )
+        """Distribution algorithm used to distribute power between batteries."""
```
```suggestion
        """The distribution algorithm used to distribute power between batteries."""
```

llucax commented on src/frequenz/sdk/microgrid/client/_client.py:
```diff
@@ -42,8 +42,8 @@
 
-# Default timeout applied to all gRPC calls
 DEFAULT_GRPC_CALL_TIMEOUT = 60.0
+"""Default timeout for gRPC calls made by this client (in seconds)."""
```
```suggestion
"""The default timeout for gRPC calls made by this client (in seconds)."""
```

llucax commented on src/frequenz/sdk/microgrid/client/_client.py:
```diff
@@ -215,7 +216,14 @@ def __init__(
+        """
+        Server (host:port) to be used for asyncio-supporting gRPC channel that
+        the client should use to contact the API.
+        """
```
Short descriptions shouldn't spawn more than one line.

```suggestion
        """The location (as "host:port") of the microgrid API gRPC server."""
```

llucax commented on src/frequenz/sdk/microgrid/client/_client.py:
```diff
@@ -215,7 +216,14 @@ def __init__(
+        """
+
         self.api = MicrogridStub(grpc_channel)
+        """gRPC stub for the microgrid API."""
```
```suggestion
        """The gRPC stub for the microgrid API."""
```

llucax commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_formula_generator.py:
```diff
@@ -57,7 +58,10 @@ class FormulaGeneratorConfig:
     """Config for formula generators."""
 
     component_ids: abc.Set[int] | None = None
+    """Component IDs to use for generating the formula."""
```
```suggestion
    """The component IDs to use for generating the formula."""
```

llucax commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_formula_generator.py:
```diff
@@ -57,7 +58,10 @@ class FormulaGeneratorConfig:
+    """Component IDs to use for generating the formula."""
+
     formula_type: FormulaType = FormulaType.PASSIVE_SIGN_CONVENTION
+    """Type of formula output."""
```
```suggestion
    """The type of formula output."""
```

llucax commented on src/frequenz/sdk/timeseries/_ringbuffer/buffer.py:
```diff
@@ -48,7 +49,10 @@ class OrderedRingBuffer(Generic[FloatArray]):
     """Time aware ringbuffer that keeps its entries sorted by time."""
 
     _DATETIME_MIN = datetime.min.replace(tzinfo=timezone.utc)
+    """The minimum datetime value."""
```
```suggestion
    """The minimum representable datetime value."""
```

llucax commented on src/frequenz/sdk/timeseries/_ringbuffer/buffer.py:
```diff
@@ -48,7 +49,10 @@ class OrderedRingBuffer(Generic[FloatArray]):
+    """The minimum datetime value."""
+
     _DATETIME_MAX = datetime.max.replace(tzinfo=timezone.utc)
+    """The maximum datetime value."""
```
```suggestion
    """The maximum representable datetime value."""
```

llucax commented on src/frequenz/sdk/timeseries/ev_charger_pool/_state_tracker.py:
```diff
@@ -22,15 +22,25 @@
+    """State of individual EV charger."""
 
     UNSPECIFIED = "UNSPECIFIED"
+    """State is unspecified."""
```
```suggestion
    """The state is unspecified."""
```

llucax commented on src/frequenz/sdk/timeseries/ev_charger_pool/_state_tracker.py:
```diff
@@ -22,15 +22,25 @@
+    """State is unspecified."""
+
     MISSING = "MISSING"
+    """EV Charger is missing."""
```
```suggestion
    """The EV Charger is missing."""
```

llucax commented on src/frequenz/sdk/timeseries/ev_charger_pool/_state_tracker.py:
```diff
@@ -22,15 +22,25 @@
+    """EV Charger is missing."""
 
     IDLE = "IDLE"
+    """EV Charger is idle."""
```
```suggestion
    """The EV Charger is idle."""
```

llucax commented on src/frequenz/sdk/timeseries/ev_charger_pool/_state_tracker.py:
```diff
@@ -22,15 +22,25 @@
+    """EV Charger is idle."""
+
     EV_PLUGGED = "EV_PLUGGED"
+    """EV Charger has an EV plugged in."""
```
```suggestion
    """The EV Charger has an EV plugged in."""
```

llucax commented on src/frequenz/sdk/timeseries/ev_charger_pool/_state_tracker.py:
```diff
@@ -22,15 +22,25 @@
+    """EV Charger has an EV plugged in."""
+
     EV_LOCKED = "EV_LOCKED"
+    """EV Charger has an EV plugged in and locked."""
```
```suggestion
    """The EV Charger has an EV plugged in and locked."""
```

llucax commented on src/frequenz/sdk/timeseries/ev_charger_pool/_state_tracker.py:
```diff
@@ -22,15 +22,25 @@
+    """EV Charger has an EV plugged in and locked."""
+
     ERROR = "ERROR"
+    """EV Charger is in an error state."""
```
```suggestion
    """The EV Charger is in an error state."""
```

llucax reviewed (commented):
Just a few more nitpicking about "The ...".

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/ev_charger_pool/_set_current_bounds.py:
```diff
@@ -24,7 +24,10 @@ class ComponentCurrentLimit:
     """A current limit, to be sent to the EV Charger."""
 
     component_id: int
+    """ID of the EV Charger."""
```
```suggestion
    """The component ID of the EV Charger."""
```

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/ev_charger_pool/_ev_charger_pool.py:
```diff
@@ -43,8 +43,13 @@ class EVChargerData:
     """Data for an EV Charger, including the 3-phase current and the component state."""
 
     component_id: int
+    """ID of the EV Charger."""
```
```suggestion
    """The component ID of the EV Charger."""
```

daniel-zullo-frequenz commented on src/frequenz/sdk/microgrid/component/_component_states.py:
```diff
@@ -15,15 +15,26 @@ class EVChargerCableState(Enum):
+    """Unspecified cable state."""
+
     UNPLUGGED = ev_charger_pb.CableState.CABLE_STATE_UNPLUGGED
+    """Cable is unplugged."""
+
     CHARGING_STATION_PLUGGED = (
         ev_charger_pb.CableState.CABLE_STATE_CHARGING_STATION_PLUGGED
     )
+    """Cable is plugged into the charging station."""
+
     CHARGING_STATION_LOCKED = (
         ev_charger_pb.CableState.CABLE_STATE_CHARGING_STATION_LOCKED
     )
+    """Cable is plugged into the charging station and locked."""
+
     EV_PLUGGED = ev_charger_pb.CableState.CABLE_STATE_EV_PLUGGED
+    """Cable is plugged into the EV."""
+
     EV_LOCKED = ev_charger_pb.CableState.CABLE_STATE_EV_LOCKED
+    """Cable is plugged into the EV and locked."""
```
```suggestion
    """The cable is unplugged."""

    CHARGING_STATION_PLUGGED = (
        ev_charger_pb.CableState.CABLE_STATE_CHARGING_STATION_PLUGGED
    )
    """The cable is plugged into the charging station."""

    CHARGING_STATION_LOCKED = (
        ev_charger_pb.CableState.CABLE_STATE_CHARGING_STATION_LOCKED
    )
    """The cable is plugged into the charging station and locked."""

    EV_PLUGGED = ev_charger_pb.CableState.CABLE_STATE_EV_PLUGGED
    """The cable is plugged into the EV."""

    EV_LOCKED = ev_charger_pb.CableState.CABLE_STATE_EV_LOCKED
    """The cable is plugged into the EV and locked."""
```

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/ev_charger_pool/_ev_charger_pool.py:
```diff
@@ -43,8 +43,13 @@ class EVChargerData:
+    """3-phase current of the EV Charger."""
+
     state: EVChargerState
+    """State of the EV Charger."""
```
```suggestion
    """The state of the EV Charger."""
```

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/ev_charger_pool/_ev_charger_pool.py:
```diff
@@ -43,8 +43,13 @@ class EVChargerData:
+    """The component ID of the EV Charger."""
+
     current: Sample3Phase[Current]
+    """3-phase current of the EV Charger."""
```
```suggestion
    """The 3-phase current of the EV Charger."""
```

daniel-zullo-frequenz commented on src/frequenz/sdk/microgrid/component/_component_states.py:
```diff
@@ -47,16 +58,33 @@ class EVChargerComponentState(Enum):
+    """Component rejected authorization."""
+
     INTERRUPTED = ev_charger_pb.ComponentState.COMPONENT_STATE_INTERRUPTED
+    """Component is interrupted."""
```
```suggestion
    """The component is starting."""

    NOT_READY = ev_charger_pb.ComponentState.COMPONENT_STATE_NOT_READY
    """The component is not ready."""

    READY = ev_charger_pb.ComponentState.COMPONENT_STATE_READY
    """The component is ready."""

    CHARGING = ev_charger_pb.ComponentState.COMPONENT_STATE_CHARGING
    """The component is charging."""

    DISCHARGING = ev_charger_pb.ComponentState.COMPONENT_STATE_DISCHARGING
    """The component is discharging."""

    ERROR = ev_charger_pb.ComponentState.COMPONENT_STATE_ERROR
    """The component is in error state."""

    AUTHORIZATION_REJECTED = (
        ev_charger_pb.ComponentState.COMPONENT_STATE_AUTHORIZATION_REJECTED
    )
    """The component rejected authorization."""

    INTERRUPTED = ev_charger_pb.ComponentState.COMPONENT_STATE_INTERRUPTED
    """The component is interrupted."""
```

daniel-zullo-frequenz commented on src/frequenz/sdk/microgrid/connection_manager.py:
```diff
@@ -141,6 +141,7 @@ async def _initialize(self) -> None:
 
 
 _CONNECTION_MANAGER: Optional[ConnectionManager] = None
+"""ConnectionManager singleton instance."""
```
```suggestion
"""The ConnectionManager singleton instance."""
```

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/ev_charger_pool/_set_current_bounds.py:
```diff
@@ -24,7 +24,10 @@ class ComponentCurrentLimit:
+    """The component ID of the EV Charger."""
+
     max_amps: float
+    """Maximum current in amps, that an EV can draw from this EV Charger."""
```
```suggestion
    """The maximum current in amps, that an EV can draw from this EV Charger."""
```

daniel-zullo-frequenz committed:
Remove PV_ARRAY component category

The PV_ARRAY component category is from the early
MicrogridData days and it is no longer.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

shsms commented:
maybe "Rename variable name" -> "Rename variable" in first commit message?

llucax commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_formula_generator.py:
```diff
@@ -31,7 +31,14 @@ class ComponentNotFound(FormulaGenerationError):
+graph to associate with it. Thus, the non-existing component ID is subscribed
+instead so that the formula engine can send `None` or `0` values at the same
+frequency as the other streams.
+"""
```
This should probably go in the commit about adding/improving attribute docs.

daniel-zullo-frequenz replied:
Indeed, it was an oversight when rebasing :(

daniel-zullo-frequenz replied:
Amended

llucax reviewed (approved):
Only one minor comment about commits structure. LGTM. Thanks!

shsms commented on tests/microgrid/test_graph.py:
```diff
@@ -1323,20 +1316,6 @@ def test__validate_leaf_components(self) -> None:
         ):
             graph._validate_leaf_components()
 
-        graph._graph.clear()
-        graph._graph.add_nodes_from(
-            [
-                (1, asdict(Component(1, ComponentCategory.GRID))),
-                (4, asdict(Component(4, ComponentCategory.EV_CHARGER))),
-                (5, asdict(Component(5, ComponentCategory.PV_ARRAY))),
-            ]
-        )
-        graph._graph.add_edges_from([(1, 5), (5, 4)])
-        with pytest.raises(
-            gr.InvalidGraphError, match="Leaf components with graph successors"
-        ):
-            graph._validate_leaf_components()
-
```
I think we should replace PV_ARRAY with Battery in this case, instead of removing it.

daniel-zullo-frequenz replied:
Ok, the test above uses the battery so we might be duplicating tests as they both validate leaf components. Anyway I can replace it with BATTERY if you think it is worth to keep it

shsms replied:
oh, that case is tested already.  didn't see that.  you can remove it if it is duplicated.

shsms replied:
I forgot that the component graph has really comprehensive tests. :-)

daniel-zullo-frequenz replied:
Then I'll keep it as it is
